File: AI_UoB/programs/data_structures_for_ai/3_stack.py
class Stack:
    def __init__(self):
        self.stack = []

    def push(self, value):
        self.stack.append(value)

    def pop(self):
        if self.isEmpty():
            print("Stack Underflow")
            return
        print(f"Popped: {self.stack.pop()}")

    def peek(self):
        if self.isEmpty():
            print("Stack is Empty")
            return
        return self.stack[-1]

    def isEmpty(self):
        return len(self.stack) == 0

    def display(self):
        if self.isEmpty():
            print("Stack is Empty")
            return
        print("TOP")
        for item in reversed(self.stack):
            print(f" {item}")
        print("BOTTOM")

if __name__ == "__main__":
    s = Stack()
    s.push(10)
    s.push(20)
    s.push(30)
    s.push(40)
    s.display()
    print(f"Peek: {s.peek()}")
    s.pop()
    s.display()
    print(f"Empty: {s.isEmpty()}")

File: AI_UoB/programs/data_structures_for_ai/3_stack.c
#include <stdio.h>
#include <stdlib.h>

#define MAX 100

int stack[MAX];
int top = -1;

void push(int value) {
    if (top == MAX - 1) {
        printf("Stack Overflow\n");
        return;
    }
    stack[++top] = value;
}

void pop() {
    if (top == -1) {
        printf("Stack Underflow\n");
        return;
    }
    printf("Popped: %d\n", stack[top--]);
}

int peek() {
    if (top == -1) {
        printf("Stack is Empty\n");
        return -1;
    }
    return stack[top];
}

int isEmpty() {
    return top == -1;
}

void display() {
    if (top == -1) {
        printf("Stack is Empty\n");
        return;
    }
    printf("TOP\n");
    for (int i = top; i >= 0; i--)
        printf(" %d\n", stack[i]);
    printf("BOTTOM\n");
}

int main() {
    push(10);
    push(20);
    push(30);
    push(40);
    display();
    printf("Peek: %d\n", peek());
    pop();
    display();
    printf("Empty: %d\n", isEmpty());
    return 0;
}